### PR TITLE
Lint paths per Coursera frontend file organization

### PR DESCRIPTION
**Pull request by turadg**
ptal @jnwng @eleith

### CODE REVIEW

File: src/lint/linter/ArcanistCourserorgLinter.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  const MISPLACED_COMPONENT_STYL = 4;
+
+  public function getInfoName() {
+    return pht('Couserorg');
+  }
+
+  public function getInfoDescription() {
+    return pht(
+      'Stifles developer creativity by requiring files to conform to '.
+      'Coursera Frontend File Organization, by edict of the FCC.');
```
Comment (eleith) [inline src/lint/linter/ArcanistCourserorgLinter.php]: is their a creativity stifling document we could link to?

File: src/lint/linter/ArcanistCourserorgLinter.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            'Bundle names must be lowercase.'));
+
+      if (!file_exists('static/nls/'.$bundleName.'.js'))
+        $this->raiseLintAtPath(
+          self::MISSING_BUNDLE_NLS,
+          pht(
+            'Every bundle with components must have '.
+            'an NLS file.'));
+
+    } else if (preg_match('/^static\/(.+)\/components/', $path)) {
```
Comment (eleith) [inline src/lint/linter/ArcanistCourserorgLinter.php]: don't we have components in open-course? or have all those moved out?

File: src/lint/linter/ArcanistCourserorgLinter.php (hunk 1 above)
Comment (turadg) [inline src/lint/linter/ArcanistCourserorgLinter.php]: I'll draft a memo and get consensus to publish

File: src/lint/linter/ArcanistCourserorgLinter.php (hunk 2 above)
Comment (turadg) [inline src/lint/linter/ArcanistCourserorgLinter.php]: we still do, yeah, but I'd like those to begin counting as linting errors. they won't block anyone, just show up highlighted in blue in the revision. people can use judgment for when to Request they move.

File: src/lint/linter/ArcanistCourserorgLinter.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      self::MISSING_BUNDLE_NLS => pht('Missing NLS for bundle'),
+      self::MISPLACED_COMPONENT => pht('Misplaced component'),
+      self::MISPLACED_COMPONENT_STYL => pht('Misplaced component styl'),
+    );
+  }
+
+  public function lintPath($path) {
+    if (preg_match('/([^_]*)__styles__\/(.+).styl/', $path, $matches)) {
+      $jsxFilename = $matches[1].$matches[2].'.jsx';
+      print_r($jsxFilename);
```
Comment (jnwng) [inline src/lint/linter/ArcanistCourserorgLinter.php]: remove print?

File: src/lint/linter/ArcanistCourserorgLinter.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      if (!file_exists($jsxFilename))
+        $this->raiseLintAtPath(
+          self::MISPLACED_COMPONENT_STYL,
+          pht(
+            'Stylus in __styles__ must correspond to a JSX in parent components.'));
```
Comment (jnwng) [inline src/lint/linter/ArcanistCourserorgLinter.php]: this is slightly strange wording, perhaps "Stylus files in __styles__ must have a corresponding React component"?

File: src/lint/linter/ArcanistCourserorgLinter.php (hunk 2 above)
Comment (eleith) [inline src/lint/linter/ArcanistCourserorgLinter.php]: @jnwng are you ok with that?

Comment (jnwng) [inline src/lint/linter/ArcanistCourserorgLinter.php]: i wasn't aware that this was a thing, what happens when we have page-specific utilizations of components? i can see there being things (some that i'm working on now) that just wrap common components but probably not ought to be shared

Comment (turadg) [inline src/lint/linter/ArcanistCourserorgLinter.php]: The WEB repo schema proposal moves all components out of pages to bundles. The current JSX localization pipeline depends on it.

**Comment by turadg** [comment]
@jnwng I cleaned up the print and help string. I also added a comment on the components in bundles idea that we've been discussing in person.

good to land?

File: src/lint/linter/ArcanistCourserorgLinter.php (hunk 2 above)
Comment (jnwng) [inline src/lint/linter/ArcanistCourserorgLinter.php]: does this mean that all our open-course components cannot currently be localized? this is a blocker

**Comment by jnwng** [comment]
:+1:

**Pull request merged**